### ISSUE TITLE
Use cross entropy loss function in the classification task.

### ISSUE THREAD

**Issue by Realvincentyuan** [google-research/bert#824]
## Background
I am doing a binary classification task with BERT and I noticed that in the run_classifier.py file the loss function was not cross entropy. So I am re-writing the loss function with cross entropy but met minor issues of the dimension.

## Codes concerning creating model
```
def custom_create_model(bert_config, is_training, input_ids, input_mask, segment_ids,labels, num_labels, use_one_hot_embeddings):
      model = modeling.BertModel(
      config=bert_config,
      is_training=is_training,
      input_ids=input_ids,
      input_mask=input_mask,
      token_type_ids=segment_ids,
      use_one_hot_embeddings=use_one_hot_embeddings)
     
    output_layer = model.get_pooled_output()

    hidden_size = output_layer.shape[-1].value

    output_weights = tf.get_variable(
      "output_weights", [num_labels, hidden_size],
      initializer=tf.truncated_normal_initializer(stddev=0.02))

    output_bias = tf.get_variable(
      "output_bias", [num_labels], initializer=tf.zeros_initializer())

    with tf.variable_scope("loss"):
        if is_training:
              # I.e., 0.1 dropout
            output_layer = tf.nn.dropout(output_layer, keep_prob=0.9)

        logits = tf.matmul(output_layer, output_weights, transpose_b=True)
        logits = tf.nn.bias_add(logits, output_bias)
        probabilities = tf.nn.softmax(logits, axis=-1)

#       Custom cross entropy loss function
        
        one_hot_labels = tf.one_hot(labels, depth=num_labels, dtype=tf.float32)
        
        print('Labels shape', tf.shape(labels))
        print('logits shape', tf.shape(logits))
        
        per_example_loss = tf.nn.softmax_cross_entropy_with_logits(labels=one_hot_labels, logits=logits)
        loss = tf.reduce_mean(per_example_loss)
        return (loss, per_example_loss, logits, probabilities)
```

## Questions
- why does BERT not use cross entropy in the classification task?
- I want to penalizIe the false positives and false negatives so I used class weight calculator in Sklearn libraries to calculate the weights and broadcast to the loss

```
from sklearn.utils.class_weight import compute_class_weight
from sklearn.utils import class_weight
class_weights = class_weight.compute_class_weight('balanced',
                                                 np.unique(df['CII'].tolist()),
                                                 df['CII'].tolist())
...# in the create_model func:
per_example_loss = tf.nn.softmax_cross_entropy_with_logits(labels=one_hot_labels, logits=logits)
weighted_losses = per_example_loss * weights
    
loss = tf.reduce_mean(weighted_losses)
```

The weights become a parameter to be tuned anyway, not sure if anybody has some tactics on the tuning?

**Comment by annous28** [comment]
Hi, I am facing the same issue. Did you manage to find a solution for this problem. I want also to use cross entropy loss for a classification problem.

**Comment by Realvincentyuan** [comment]
> Hi, I am facing the same issue. Did you manage to find a solution for this problem. I want also to use cross entropy loss for a classification problem.

Yes, I just replaced the default loss function with a weighted cross entropy. What problems did you meet?

**Comment by InhyeokYoo** [comment]
NLL loss is just the same as the cross-entropy loss. So actually the model uses cross-entropy.

One thing I want to ask is that is the weighted CE better than focal loss? My dataset is extremely imbalanced so I need to penalty incorrect classification of the minor class dataset.
